### DISCUSSION TITLE
Versioning models

### DISCUSSION THREAD

**Discussion by helloscoopa** [Automattic/mongoose#11508] [Q&A] [answered]
This might be a noob question, but I couldn't find a proper solution to overcome this at least for the architecture that I was about to build. So my question is, in my NodeJS backend I need to run multiple versions in the same repo. so what I did is started creating new folders for the versions.

models/
   v1/user.js
   v2/user.js

something like this. and I'm importing both of these modules in my startup scripts so that they can be used interchangeably.

**v1/user.js**

```
const mongoose = require("mongoose")

const userSchema = new mongoose.Schema({
    name: { type: String, required: true }
})

const User = mongoose.model("User", userSchema)

exports.User = User
```

**v2/user.js**

```
const mongoose = require("mongoose")

const userSchema = new mongoose.Schema({
    name: { type: String, required: true },
    new_field: { type: String, required: true }
})

const User = mongoose.model("User", userSchema)

exports.User = User
```

So in this context, I'm getting the error '_Cannot overwrite User model once compiled._' which makes perfect sense as I use the same name in both occurrences. I just need to know if there is a way to eliminate it. because at the end of the day we refer to what model we need to use in each transaction right so I feel just eliminating the error with some trick is theoretically doable.

or else please point me out a way to do this in another way, I need to keep the initial version of the model anyways because the v1 routes use that schema already where I can't randomly add more required fields into it.

Any help regarding this is highly appreciated.

**Comment by 1finedev** [comment]
Mind if i ask, are you using the same database for both models?

**Reply by helloscoopa** [reply]
Yes, I am. I'm trying to do a model upgrade to the existing API without breaking the existing endpoints. As newly added props are required, I can't just add them into the old one. Also, the related endpoints will be depreciated once I write the new v2 endpoints. but between the switch (until the client apps get updated) I need both v1 and v2 endpoints to work which means I need to have both v1 and v2 models.

**Comment by 1finedev** [comment, marked as answer]
but if it helps you can try either of this approach 

```js
//add this line before exporting your model to empty the models each time 
mongoose.models = {}; 
const User = mongoose.model("User", userSchema)
```

or try this

```js
export const User = mongoose.models.User || mongoose.model('User', userSchema);
```

**Reply by helloscoopa** [reply]
Thanks, however doing `mongoose.models = {}` is not possible and `models` is not an assignable at least in the `mongoose v5.13.5`. but as it looks promising, I was looking for a way to properly remove a model which is already registered and found [this](https://stackoverflow.com/questions/19643126/how-do-you-remove-a-model-from-mongoose).

So using `delete mongoose.connection.models['User']`, I can remove existing model compilations. So what I'm doing now is, before initializing each model, I clear the User model so that I clear if another version ever complied before. This worked out pretty fine for my requirement. 

Thanks for your contribution, Cheers!

**Reply by 1finedev** [reply]
Cheers
